serverlessgo/main.go

```diff
@@ -60,14 +60,33 @@ func (lb *loadBalancer) terminate() {
 		i.terminate()
 	 }
 	lb.isTerminated = true
-	i.cond.Set(true)
+	lb.arrivalCond.Set(true)
 }
 
 func (lb *loadBalancer) nextInstance() *instance {
-	instance := newInstance(len(lb.instances))
-	lb.instances = append(lb.instances, instance)
-	godes.AddRunner(instance)
-	return instance
+	var instanceChoosed *instance
+	var instanceIndex int
+	hasChosen := false
+	for i := 0; i < len(lb.instances); i++ {
+		instance := lb.instances[i]
+		if !instance.isWorking() {
+			instanceChoosed = instance
+			instanceIndex = i
+			hasChosen = true
+			break
+		}
+	}
+
+	if hasChosen {
+		// removes the chosen instance from the array
+		lb.instances = append(lb.instances[:instanceIndex], lb.instances[instanceIndex+1:]...)
```

Comment:
lb.instances só deve ser removido quando for a hora de fazer scale down. Acho que você está misturando conceitos. Você deve remover possibilidades associadas a requisições e isso vem depois.

Reply:
Não. O que acontece é que quando eu seleciono uma instância, eu a removo e a adiciono na primeira posição do array (logo abaixo do else), pois a maneira que seleciono instâncias dentre as instâncias existentes privilegia as instâncias mais no início do array.

Reply:
Acho que o jeito mais claro de fazer o que você está querendo é que [sort](https://golang.org/pkg/sort/). Você deve transformar um slice de instâncias que o lb manipula num slice ordenável. Daí vai ficar super claro que você quer ordenar e quais os critérios para ordenação.

Dá uma olhada e me fala alguma coisa.